src/Soil-Core/SoilMaterializer.class.st

```diff
@@ -37,16 +37,28 @@ SoilMaterializer >> materializeFromBytes: aByteArray [
 
 { #category : #'instance creation' }
 SoilMaterializer >> newObject [
-	| objectClass description objectIndex|
+	| objectClass description objectIndex versions obj |
 	objectIndex := self nextLengthEncodedInteger.
 	description := objectIndex isZero
 		ifTrue: [ SOBehaviorDescription meta ]
 		ifFalse: [
 			transaction 
 				behaviorDescriptionWithId: (externalObjectRegistry basicReferenceAt: objectIndex) 
 				ifNone: [ self halt  ] ].
-	objectClass := description objectClass.
-	^ objectClass classLayout soilBasicMaterialize: objectClass with: self
+		objectClass := description objectClass.
+	^ description isCurrent 
+		ifTrue: [ 
+			objectClass classLayout soilBasicMaterialize: objectClass with: self ]
+		ifFalse: [ 
+			versions := transaction behaviorVersionsUpTo: description.
+			obj := objectClass basicNew.
+			self registerObject: obj.
+			versions last instVarNames do: [ :instVar | | value |
+				value := self nextSoilObject.
+				(versions allSatisfy: [ :version | version instVarNames includes: instVar ]) ifTrue: [ 
```

Comment:
Do I understand this correctly?
the allSatisfy: check is needed as if the var would be removed at some point int the past, it should be nil, even though there is a new var with the same name.